indico/modules/categories/forms.py

```diff
@@ -65,12 +69,34 @@ class CategorySettingsForm(IndicoForm):
     notify_managers = BooleanField(_('Notify managers'), widget=SwitchWidget(),
                                    description=_('Whether to send email notifications to all managers of this category '
                                                  'when an event is created inside it or in any of its subcategories.'))
+    google_wallet_enabled = BooleanField(_('Google Wallet'), widget=SwitchWidget(),
+                                         description=_('Add Google Wallet configuration.'))
+    google_wallet_application_credentials = TextAreaField(_('Google Wallet application credentials'),
+                                                          [DataRequired(),
+                                                           HiddenUnless('google_wallet_enabled', preserve_data=True)],
+                                                          description=_('Json content generated from Google Wallet '
+                                                                        'console.'))
+    google_wallet_issuer_name = StringField(_('Google Wallet Issuer Name'),
+                                            [DataRequired(),
+                                             HiddenUnless('google_wallet_enabled', preserve_data=True)],
+                                            description=_('Issuer name that will appear in the Google Wallet ticket '
+                                                          'top header.'))
+    google_wallet_issuer_id = StringField(_('Google Wallet Issuer ID'),
+                                          [DataRequired(),
+                                           HiddenUnless('google_wallet_enabled', preserve_data=True)],
+                                          description=_('Issuer ID assigned in the Google Wallet console.'))
+
     event_creation_notification_emails = EmailListField(_('Notification E-mails'),
                                                         description=_('List of emails that will receive a notification '
                                                                       'every time a new event is created inside the '
                                                                       'category or one of its subcategories. '
                                                                       'One email address per line.'))
 
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        if not config.ENABLE_GOOGLE_WALLET:
+            self.google_wallet_enabled.widget = HiddenInput()
```

Comment:
This is creating issues: the google_wallet_enabled field is included in the BASIC_FIELDS tuple, if I remove it then I get an error when trying to render the form.
I suggest to keep the HiddenInput widget change.
What do you suggest?

Reply:
So... I'm testing this:

```
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        if not config.ENABLE_GOOGLE_WALLET:
            del self.google_wallet_enabled
            self.BASIC_FIELDS = tuple(field for field in self.BASIC_FIELDS if field != 'google_wallet_enabled')
```

...but still getting this error when ENABLE_GOOGLE_WALLET is set to False:

KeyError: 'google_wallet_enabled'

Reply:
got a traceback?

Reply:
Full Traceback:

File "/Users/giorgiopieretti/Work/indico-vanilla/vanilla-env/lib/python3.9/site-packages/sentry_sdk/integrations/flask.py", line 89, in sentry_patched_wsgi_app
return SentryWsgiMiddleware(lambda *a, **kw: old_app(self, *a, **kw))(
File "/Users/giorgiopieretti/Work/indico-vanilla/vanilla-env/lib/python3.9/site-packages/sentry_sdk/integrations/wsgi.py", line 115, in __call__
reraise(*_capture_exception(hub))
File "/Users/giorgiopieretti/Work/indico-vanilla/vanilla-env/lib/python3.9/site-packages/sentry_sdk/_compat.py", line 115, in reraise
raise value
File "/Users/giorgiopieretti/Work/indico-vanilla/vanilla-env/lib/python3.9/site-packages/sentry_sdk/integrations/wsgi.py", line 108, in __call__
rv = self.app(
File "/Users/giorgiopieretti/Work/indico-vanilla/vanilla-env/lib/python3.9/site-packages/sentry_sdk/integrations/flask.py", line 89, in <lambda>
return SentryWsgiMiddleware(lambda *a, **kw: old_app(self, *a, **kw))(
File "/Users/giorgiopieretti/Work/indico-vanilla/vanilla-env/lib/python3.9/site-packages/flask/app.py", line 1478, in __call__
return self.wsgi_app(environ, start_response)
File "/Users/giorgiopieretti/Work/indico-vanilla/vanilla-env/lib/python3.9/site-packages/flask/app.py", line 1458, in wsgi_app
response = self.handle_exception(e)
File "/Users/giorgiopieretti/Work/indico-vanilla/vanilla-env/lib/python3.9/site-packages/flask/app.py", line 1455, in wsgi_app
response = self.full_dispatch_request()
File "/Users/giorgiopieretti/Work/indico-vanilla/vanilla-env/lib/python3.9/site-packages/flask/app.py", line 869, in full_dispatch_request
rv = self.handle_user_exception(e)
File "/Users/giorgiopieretti/Work/indico-vanilla/vanilla-env/lib/python3.9/site-packages/flask/app.py", line 759, in handle_user_exception
return self.ensure_sync(handler)(e)
File "/Users/giorgiopieretti/Work/indico-vanilla/vanilla-env/lib/python3.9/site-packages/flask/app.py", line 867, in full_dispatch_request
rv = self.dispatch_request()
File "/Users/giorgiopieretti/Work/indico-vanilla/vanilla-env/lib/python3.9/site-packages/flask/app.py", line 852, in dispatch_request
return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)
File "/Users/giorgiopieretti/Work/indico-vanilla/src/indico/web/flask/util.py", line 80, in wrapper
return obj().process()
File "/Users/giorgiopieretti/Work/indico-vanilla/src/indico/web/rh.py", line 292, in process
res = self._do_process()
File "/Users/giorgiopieretti/Work/indico-vanilla/src/indico/web/rh.py", line 263, in _do_process
rv = self._process()
File "/Users/giorgiopieretti/Work/indico-vanilla/src/indico/modules/categories/controllers/management.py", line 146, in _process
return WPCategoryManagement.render_template('management/settings.html', self.category, 'settings', form=form,
File "/Users/giorgiopieretti/Work/indico-vanilla/src/indico/web/views.py", line 129, in render_template
return cls(g.rh, *wp_args, **context).display()
File "/Users/giorgiopieretti/Work/indico-vanilla/src/indico/web/views.py", line 276, in display
body = self._display(params)
File "/Users/giorgiopieretti/Work/indico-vanilla/src/indico/web/views.py", line 365, in _display
return self._apply_decoration(self._get_body(params | self._kwargs))
File "/Users/giorgiopieretti/Work/indico-vanilla/src/indico/modules/categories/views.py", line 45, in _get_body
return self._get_page_content(params)
File "/Users/giorgiopieretti/Work/indico-vanilla/src/indico/web/views.py", line 159, in _get_page_content
return self.render_template_func(template, **params)
File "/Users/giorgiopieretti/Work/indico-vanilla/vanilla-env/lib/python3.9/site-packages/flask/templating.py", line 152, in render_template
return _render(app, template, context)
File "/Users/giorgiopieretti/Work/indico-vanilla/vanilla-env/lib/python3.9/site-packages/flask/templating.py", line 133, in _render
rv = template.render(context)
File "/Users/giorgiopieretti/Work/indico-vanilla/vanilla-env/lib/python3.9/site-packages/jinja2/environment.py", line 1301, in render
self.environment.handle_exception()
File "/Users/giorgiopieretti/Work/indico-vanilla/vanilla-env/lib/python3.9/site-packages/jinja2/environment.py", line 936, in handle_exception
raise rewrite_traceback_stack(source=source)
File "/Users/giorgiopieretti/Work/indico-vanilla/src/indico/modules/categories/templates/management/settings.html", line 10, in top-level template code
{% call form_footer(image_form, skip_label=true) %}
File "/Users/giorgiopieretti/Work/indico-vanilla/src/indico/modules/categories/templates/management/base.html", line 31, in top-level template code
data-category-id="{{ category.id }}"></a>
File "/Users/giorgiopieretti/Work/indico-vanilla/src/indico/web/templates/layout/management_page.html", line 4, in top-level template code
{% set show_banner_area = self.display_view_button() or self.banner_title() or self.banner_actions() %}
File "/Users/giorgiopieretti/Work/indico-vanilla/src/indico/web/templates/layout/base.html", line 1, in top-level template code
{% block page %}
File "/Users/giorgiopieretti/Work/indico-vanilla/src/indico/web/templates/layout/management_page.html", line 39, in block 'page'
{{ super() }}
File "/Users/giorgiopieretti/Work/indico-vanilla/src/indico/web/templates/layout/base.html", line 51, in block 'page'
{% block content %}{% endblock %}
File "/Users/giorgiopieretti/Work/indico-vanilla/src/indico/modules/categories/templates/management/settings.html", line 30, in block 'content'
{{ form_rows(form, fields=form.BASIC_FIELDS) }}
File "/Users/giorgiopieretti/Work/indico-vanilla/vanilla-env/lib/python3.9/site-packages/jinja2/runtime.py", line 777, in _invoke
rv = self._func(*arguments)
File "/Users/giorgiopieretti/Work/indico-vanilla/src/indico/web/templates/forms/_form.html", line 98, in template
{{ form_row(field, widget_attrs.get(field.short_name, {}), orientation, skip_label=skip_labels,
File "/Users/giorgiopieretti/Work/indico-vanilla/vanilla-env/lib/python3.9/site-packages/jinja2/runtime.py", line 777, in _invoke
rv = self._func(*arguments)
File "/Users/giorgiopieretti/Work/indico-vanilla/src/indico/web/templates/forms/_form.html", line 90, in template
{{ form_field(field, classes, widget_attrs, disabled=disabled, hide_description=hide_description) }}
File "/Users/giorgiopieretti/Work/indico-vanilla/vanilla-env/lib/python3.9/site-packages/jinja2/runtime.py", line 777, in _invoke
rv = self._func(*arguments)
File "/Users/giorgiopieretti/Work/indico-vanilla/src/indico/web/templates/forms/_form.html", line 47, in template
{{ _render_field(field, widget_attrs, disabled) }}
File "/Users/giorgiopieretti/Work/indico-vanilla/src/indico/web/forms/jinja_helpers.py", line 85, in render_field
args = _attrs_for_validators(field, field.validators)
File "/Users/giorgiopieretti/Work/indico-vanilla/src/indico/web/forms/jinja_helpers.py", line 65, in _attrs_for_validators
condition_field = field.get_form()[validator.field]
File "/Users/giorgiopieretti/Work/indico-vanilla/vanilla-env/lib/python3.9/site-packages/wtforms/form.py", line 63, in __getitem__
return self._fields[name]
KeyError: 'google_wallet_enabled'

Reply:
I'm pretty sure this is happening because you do not delete all the google-wallet-related fields so the other fields still reference this one in via their `HiddenUnless` validators

Reply:
You are right!
I found a way to fix it... it works now, tnx :)